url7a45f192ca_mfullen_ping_pong_net_tgz-pJ8-ping_pong_net_client_io_TestClientNonPPNServerJ8/src/ping/pong/net/connection/io/AbstractIoConnection.java

```diff
@@ -0,0 +1,475 @@
+    }
+
+    @Override
+    public synchronized void close()
+    {
+        if (!this.canStart)
+        {
+            LOGGER.warn("Connection cannot be closed, it never started");
+            return;
+        }
+        if (this.closed)
+        {
+            LOGGER.warn("Connection is already Closed");
+            return;
+        }
+
+        this.connected = false;
+        if (this.isAnyRunning())
+        {
+            this.fireOnSocketClosed();
+            disconnect();
+
+            if (this.isTcpWriteRunning())
+            {
+                this.ioTcpWriteRunnable.close();
+            }
+
+            if (this.isTcpReadRunning())
+            {
+                this.ioTcpReadRunnable.close();
+            }
+
+            this.closeUdp();
+        }
+
+        LOGGER.info("Connection ({}) has been closed", this.getConnectionId());
+        this.closed = true;
+    }
+
+    @Override
+    public synchronized boolean isConnected()
+    {
+        return this.connected;
+    }
+
+    @Override
+    public synchronized int getConnectionId()
+    {
+        return this.connectionId;
+    }
+
+    @Override
+    public synchronized void setConnectionId(int connectionId)
+    {
+        this.connectionId = connectionId;
+    }
+
+    @Override
+    public void sendMessage(Envelope<MessageType> message)
+    {
+        this.enqueueMessageToWrite(message);
+        LOGGER.debug("Preparing to send Message {}", message);
+    }
+
+    /**
+     * This method enqueues a message on the TcpWrite thread for sending
+     *
+     * @param msg the message to enqueue for sending
+     */
+    protected void sendTcpMessage(MessageType msg)
+    {
+        if (this.ioTcpWriteRunnable != null)
+        {
+            LOGGER.trace("Enqueued {} TCP Message to write", msg);
+            boolean enqueueMessage = this.ioTcpWriteRunnable.enqueueMessage(msg);
+            LOGGER.trace("Message Enqueued {}", enqueueMessage);
+        }
+        else
+        {
+            LOGGER.trace("IoTcpWrite is null");
+        }
+    }
+
+    @Override
+    public boolean isUdpEnabled()
+    {
+        return udpEnabled;
+    }
+
+    @Override
+    public void enqueueReceivedMessage(MessageType message)
+    {
+        boolean add = this.receiveQueue.add(message);
+        LOGGER.trace("Enqueued message {}", add);
+    }
+
+    @Override
+    public void enqueueMessageToWrite(Envelope<MessageType> message)
+    {
+        if (message.isReliable())
+        {
+            sendTcpMessage(message.getMessage());
+        }
+        else
+        {
+            sendUdpMessage(message.getMessage());
+        }
+    }
+
+    @Override
+    public void addConnectionEventListener(ConnectionEvent listener)
+    {
+        this.connectionEventListeners.add(listener);
+    }
+
+    @Override
+    public void removeConnectionEventListener(ConnectionEvent listener)
+    {
+        this.connectionEventListeners.remove(listener);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **473**. At this line the following field is not protected by synchronization: `connectionEventListeners`. Your code typically protects this field with synchronization on object: `AbstractIoConnection.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `AbstractIoConnection.this`. For example, like at lines: **268**, **245**, and **257**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **245**, **257**, and **268**.